hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineManager.java

```diff
@@ -641,13 +663,26 @@ public boolean getSafeModeStatus() {
   }
 
   @Override
-  public void handleSafeModeTransition(
+  public synchronized void handleSafeModeTransition(
       SCMSafeModeManager.SafeModeStatus status) {
-    this.isInSafeMode.set(status.getSafeModeStatus());
-    if (!status.getSafeModeStatus()) {
-      // TODO: #CLUTIL if we reenter safe mode the fixed interval pipeline
-      // creation job needs to stop
+    // TODO: #CLUTIL - handle safemode getting re-enabled
+    boolean currentAllowPipelines =
+        allowPipelineCreation.getAndSet(status.isPreCheckComplete());
+    boolean currentlyInSafeMode =
+        isInSafeMode.getAndSet(status.isInSafeMode());
+    boolean triggerPipelines = false;
+
+    // Trigger pipeline creation only if the preCheck status has changed to
+    // complete.
+    if (allowPipelineCreation.get() && !currentAllowPipelines) {
+      triggerPipelines = true;
+    }
+    // Start the pipeline creation thread only when safemode switches off
+    if (!getSafeModeStatus() && currentlyInSafeMode) {
       startPipelineCreator();
+      triggerPipelines = true;
+    }
+    if (triggerPipelines) {
```

Comment:
Let me understand a bit this part of the code:
we trigger pipelines in two cases:
1: the pipeline creation was not allowed before, but now it is allowed, effectively when precheck is complete
2: safemode is transitioning to off, and currently we are in safe mode

Looking at the code of triggerPipelineCreation, it is calling the same method from the BackgroundPipelineCreator, which is doing this only after the creator itself is started.
Now this means that we are effectively doing the same as the old code, and trigger pipeline creation after leaving safe mode, am I understand this well?

Reply:
In the logic prior to this patch, `BackgroundPipelineCreator` only gets started when safemode exits. However, the creator does not need to be started for pipelines to get created. Infact, in order to exit safemode, there must be a replication factor 3 pipeline created, and that must happen without the creator running.

There are actually two ways a pipeline can be created:

1. When the backgroundPipelineCreator is running, it starts every 2 minutes and tries to create pipelines.

2. When a node registers with SCM, it triggers an event and calls `backgroundPipelineCreator.triggerPipelineCreation();` which tries to create pipelines immediately - even if the BackgroundPipelineCreator has already been started.

Therefore in safemode, before this patch, pipelines are initially created by (2) above. However now we want to stop that happening. Therefore whenever the precheck completes (enough nodes registered) we trigger the pipeline creation, but hold off on starting the background creator thread. Then it is started when safemode exits.

Now that you have highlighted this, it is possibly valid to just start the BackgroundPipeLineCreator in both cases. However there is one further detail - in the existing implementation, there is a delay for safemode exiting. Once the rules have passed some services are not started for a delay of 5 minutes and the BackgroundPipelineCreator thread is one of them. I am not sure on the history of that so it probably makes sense to keep that logic for now.

Reply:
Oh, I missed one small detail in the shouldSchedulePipelineCreator method... and misinterpreted it sorry about that... So triggerPipelineCreation will create pipelines if the safemode precheck is finished and allowPipelineCreation is already true in SCMPipelineManager.

With that it makes perfect sense, that we want to trigger the pipeline creation once when precheck is complete, and start the scheduled runs after the safe mode is exited.

But in this case, it would be enough to call triggerPipelineCreation when we arrived to the end of the precheck stage, and start the background pipeline creator schedule when we left safe mode.

Now what I understand happens at a startup when this is interesting is the following:
for (0..nodecount) {
  triggerPipelineCreation() // noop as allowPipelineCreation is false
}
precheck finished
triggerPipelineCreation // runs and creates pipelines
safe mode exit
startFixedIntervalPipelineCreator // calls createPipelines right away once at start
triggerPipelineCreation // tries to call createPipelines if it is not running from the prev. call

I think we don't need this last call, which makes the triggerPipelines boolean unnecessary and the code a bit simpler. Is it correct, or I am still missing something? :)

Reply:
Good observation - you are correct. I have simplified this as you suggested.